### PR TITLE
WIP : speeding up syncing

### LINKED ISSUE #509: Memoize timestamp fetching in syncer
Following line contributes to slowness of syncing. 

https://github.com/district0x/memefactory/blob/fc30420e1297c015fb69a3cf1cccca4667c5c7ea/src/memefactory/server/syncer.cljs#L373

Use clojure's [memoize](https://clojuredocs.org/clojure.core/memoize) function for caching timestamp for particular block number. In MF there's many events with the same block number, because many cards are transferred at once.

### PR DESCRIPTION
**Pull request by fbielejec**
closes #509

### CODE REVIEW

File: src/memefactory/server/syncer.cljs [hunk 1]
```diff
@@ -310,45 +310,60 @@
...
+   (fn [resolve reject]
+     (web3-eth/get-block @web3 block-number false (fn [err {:keys [:timestamp] :as res}]
+                                                    (if err
+                                                      (reject err)
+                                                      (do
+                                                        (log/debug "cache miss for block-timestamp" {:block-number block-number :timestamp timestamp})
+                                                        (resolve timestamp))))))))
+
+(def block-timestamp
+  (memoize block-timestamp*))
```
Comment (madvas) [inline src/memefactory/server/syncer.cljs]: I'm not sure if memoize function is gonna actually work with promises like this

File: src/memefactory/server/syncer.cljs [hunk 2]
```diff
@@ -310,45 +310,60 @@
...
+                            :last-block-number last-block-number
+                            :last-log-index last-log-index
+                            :block-number block-number
+                            :log-index log-index}]
+                   (if (or (> block-number last-block-number)
+                           (and (= block-number last-block-number) (> log-index last-log-index)))
+                     (do
+                       (js/Promise.
+                        (fn [resolve reject]
+                          (log/info "Handling new event" evt)
```
Comment (madvas) [inline src/memefactory/server/syncer.cljs]: shouldn't this be debug? Event dispatches are already logged with info in web3-events

File: src/memefactory/server/syncer.cljs [hunk 3]
```diff
@@ -310,45 +310,60 @@
...
+                                     #(db/upsert-event! {:event/last-log-index log-index
+                                                         :event/last-block-number block-number
+                                                         :event/count (inc count)
+                                                         :event/event-name event-name
+                                                         :event/contract-key contract-key})
+                                     #(resolve ::processed)))))
+                     (do
+                       (js/Promise.
+                        (fn [resolve reject]
+                          (log/info "Skipping handling of a persisted event" evt)
```
Comment (madvas) [inline src/memefactory/server/syncer.cljs]: shouldn't this be debug? Event dispatches are already logged with info in web3-events

File: src/memefactory/server/syncer.cljs (hunk 1 above)
Comment (fbielejec) [reply src/memefactory/server/syncer.cljs]: why not? It just builds a cache, what this cache returns is secondary. Just try it ;)

File: src/memefactory/server/syncer.cljs (hunk 2 above)
Comment (fbielejec) [reply src/memefactory/server/syncer.cljs]: I'd argue these are more useful (what the system does with the raw event + filtered information. We can supress web3-event logging with a config flag, I'll add that here.

File: src/memefactory/server/syncer.cljs (hunk 1 above)
Comment (madvas) [reply src/memefactory/server/syncer.cljs]: hmm, how is what it returns secondary? 
https://clojuredocs.org/clojure.core/memoize
> The memoized version of the function keeps a cache of the mapping from arguments
to results

File: src/memefactory/server/syncer.cljs (hunk 3 above)
Comment (fbielejec) [reply src/memefactory/server/syncer.cljs]: See above

File: src/memefactory/server/syncer.cljs (hunk 2 above)
Comment (madvas) [reply src/memefactory/server/syncer.cljs]: there already is option for that `:disable-dispatch-logging?`

File: src/memefactory/server/syncer.cljs (hunk 1 above)
Comment (fbielejec) [reply src/memefactory/server/syncer.cljs]: Secondary in the sense that the result is data, any data, a number, a function, a js/object etc.

File: src/memefactory/server/syncer.cljs (hunk 2 above)
Comment (fbielejec) [reply src/memefactory/server/syncer.cljs]: In web3-events, not here

File: src/memefactory/server/syncer.cljs (hunk 1 above)
Comment (madvas) [reply src/memefactory/server/syncer.cljs]: ah, so it returns the same Promise object, that's basically already resolved?

File: src/memefactory/server/syncer.cljs (hunk 2 above)
Comment (madvas) [reply src/memefactory/server/syncer.cljs]: yeah, I mean, you can disable logging in web3-events and keep it here then

File: src/memefactory/server/syncer.cljs (hunk 1 above)
Comment (fbielejec) [reply src/memefactory/server/syncer.cljs]: Voila, i actually suspect that beneath its just building a hashmap

**Pull request merged**